File: C++/0240_searcha2dmatrixii.cc
// 2020-08-04 01:34:45

#include <vector>

using std::vector;

class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        if (matrix.size() == 0 || matrix[0].size() == 0)
            return false;
        int i = 0, j = matrix[0].size() - 1;
        while (i <= matrix.size() - 1 && j >= 0)
        {
            if (target > matrix[i][j])
                i += 1;
            else if (target < matrix[i][j])
                j -= 1;
            else
                return true;
        }
        return false;
    }
};

File: Java/0240_Search_a_2D_Matrix_II.java
// 2020-08-04 00:00:42

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0
            || matrix[0] == null || matrix[0].length == 0) {
                return false;
            }      
        int m = matrix.length, n = matrix[0].length;
        int i = 0, j = n - 1;
        while (i <= m - 1 && j >= 0) {
            if (target < matrix[i][j]) --j;
            else if (target > matrix[i][j]) ++i;
            else return true;
        }
        return false;
    }
}